cmake/kokkos_arch.cmake

```diff
@@ -389,7 +396,34 @@ IF (KOKKOS_ENABLE_OPENMPTARGET)
 ENDIF()
 
 IF(KOKKOS_ENABLE_CUDA AND NOT CUDA_ARCH_ALREADY_SPECIFIED)
-  MESSAGE(SEND_ERROR "CUDA enabled but no NVIDIA GPU architecture currently enabled.  Please give one -DKokkos_ARCH_{..}=ON' to enable an NVIDIA GPU architecture.")
+  # Try to autodetect the CUDA Compute Capability by asking the device
+  SET(_binary_test_dir ${CMAKE_CURRENT_BINARY_DIR}/cmake/compile_tests/CUDAComputeCapabilityWorkdir)
+  FILE(REMOVE_RECURSE ${_binary_test_dir})
+  FILE(MAKE_DIRECTORY ${_binary_test_dir})
+
+  EXECUTE_PROCESS(
+    COMMAND ${CMAKE_CXX_COMPILER}
+      ${CMAKE_CURRENT_SOURCE_DIR}/cmake/compile_tests/cuda_compute_capability.cu
+      -o cuda_compute_capability
+    WORKING_DIRECTORY ${_binary_test_dir}
+    OUTPUT_QUIET
+    ERROR_QUIET)
+  EXECUTE_PROCESS(
+    COMMAND ${_binary_test_dir}/cuda_compute_capability
+    RESULT_VARIABLE _result
+    OUTPUT_VARIABLE CUDA_COMPUTE_CAPABILITY)
+  IF(${_result} EQUAL 0)
+    LIST(FIND KOKKOS_CUDA_ARCH_FLAGS sm_${CUDA_COMPUTE_CAPABILITY} FLAG_INDEX)
+    IF(FLAG_INDEX EQUAL -1)
+      MESSAGE(SEND_ERROR "Unsupported compute capability ${CUDA_COMPUTE_CAPABILITY}.")
+    ELSE()
+      MESSAGE(STATUS "Detected CUDA Compute Capability ${CUDA_COMPUTE_CAPABILITY}")
+      LIST(GET KOKKOS_CUDA_ARCH_LIST ${FLAG_INDEX} ARCHITECTURE)
+      KOKKOS_SET_OPTION(${ARCHITECTURE} ON)
```

Comment:
Why did you introduce that function?  Is there an issue with doing
```CMake
SET(KOKKOS_ARCH_${ARCHITECTURE} ON)
```

Reply:
The main objective of the function is to set `Kokkos_ARCH_$ARCHITECTURE` so that it is user-visible what we detected. Since it is a `CACHE` variable, we need to `FORCE` it and overwrite all the previous settings for the option (including the docstring).

Reply:
So I have a couple of functions that sort fixes this issue. Whenever you are potentially changing a variable locally, you do `add_feature(<OPTION> <DOCSTRING>)` in CMake and then you just print out those values at the end of CMake. 

```console
-- The following features are defined/enabled (+):
     TIMEMORY_USE_CALIPER: Enable Caliper
     TIMEMORY_USE_CUDA: Enable CUDA option for GPU measurements
     TIMEMORY_USE_DYNINST: Enable dynamic instrumentation

-- The following features are NOT defined/enabled (-):
     TIMEMORY_USE_ALLINEA_MAP: Enable control for AllineaMAP sampler
     TIMEMORY_USE_ARCH: Enable architecture flags
     TIMEMORY_USE_CLANG_TIDY: Enable running clang-tidy
```

Reply:
I assume you mean `add_feature_info()`

Reply:
I think it's https://github.com/NERSC/timemory/blob/5c0f89880db25e31a0778d8252b05c4ad5e50803/cmake/Modules/MacroUtilities.cmake#L690-L707.

For me, this feels a little out-of-scope of this pull request. In the end, I care about what is in the cache. Of course, we can discuss printing a summary of the enabled and disabled features at the end of the `CMake` run separately.